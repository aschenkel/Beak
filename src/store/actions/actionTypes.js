export const SET_USER = 'SET_USER';
export const SET_PROFILE = 'SET_PROFILE';
export const SET_TWEETS = 'SET_TWEETS';
export const ERROR_USER_LOGIN = 'ERROR_USER_LOGIN';
export const LOG_OUT = 'LOG_OUT';
export const INIT_COMPLETED = 'INIT_COMPLETED';