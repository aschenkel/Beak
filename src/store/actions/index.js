import * as actionTypes from './actionTypes'
import FabricTwitterKit from 'react-native-fabric-twitterkit'
import { AsyncStorage } from 'react-native';
var Twitter = require('twitter-node-client').Twitter;
var config = require('../../../twitter_config.json');

export const setUser = (session) => ({
    type: actionTypes.SET_USER,
    userName: session.userName,
    userID: session.userID,
    authorized:true,
    authToken: session.authToken,
    authTokenSecret: session.authTokenSecret,
    error:false
})

export const signIn = () => {
    return function (dispatch) {
        return FabricTwitterKit.login((error, res) => 
        {
            if(res !== undefined){
                if(res.userName){
                    dispatch(setUser(res))
                    AsyncStorage.setItem('session', JSON.stringify(res)) 
                }else{
                    dispatch(errorLoadingUser()); 
                }
            }
        });
    };
}

export const loadUser = () => {
    return function (dispatch) {
        AsyncStorage.getItem('session').then((session) => {
            if (session) {
                dispatch(setUser(JSON.parse(session)))
            }
        })
        dispatch(initCompleted())
    };
}

export const fetchTweets = () => {
    return function (dispatch,getState) {
        var twitter = new Twitter(config);
        var error = (err, response, body) =>{
            //dispatch error
        };
        var success =  response => {
            var data = JSON.parse(response)
            dispatch(setTweets(data))
        };    
        twitter.getUserTimeline({ user_id: ""+ getState().user.userID, count: '30',include_rts:'false'}, error, success)
    };
}

export const logOut = () => {
    return function (dispatch,getState) {
        FabricTwitterKit.logOut() //Doesnt work. Issue: https://github.com/tkporter/react-native-fabric-twitterkit/issues/33
        AsyncStorage.removeItem('session')
        dispatch(logOutAction())
    };
}

export const fetchProfile= () => {
    return function (dispatch) {
        FabricTwitterKit.fetchProfile((error, profile) => 
            {
                if(profile !== undefined){
                    dispatch(setProfile(profile))
                }
            })
    };
}

export const setProfile = (profile) => ({
    type: actionTypes.SET_PROFILE,
    name: profile.name,
    imageURL: profile.profile_image_url_https
})

export const logOutAction = () => ({
    type: actionTypes.LOG_OUT,
})

export const setTweets = (tweets) => ({
    type: actionTypes.SET_TWEETS,
    tweets: tweets,
})

export const errorLoadingUser = () => ({
  type: actionTypes.ERROR_USER_LOGIN,
  error:true
})

export const initCompleted = () => ({
  type: actionTypes.INIT_COMPLETED,
  init:true
})

