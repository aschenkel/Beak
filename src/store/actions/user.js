import * as actionTypes from './actionTypes'
import FabricTwitterKit from 'react-native-fabric-twitterkit'
import { AsyncStorage } from 'react-native';

export const setUser = (session) => ({
    type: actionTypes.SET_USER,
    userName: session.userName,
    userID: session.userID,
    authorized:true,
    authToken: session.authToken,
    authTokenSecret: session.authTokenSecret,
    error:false
})

export const signIn = (api = FabricTwitterKit) => {
    return function (dispatch) {
        return api.login((error, res) => 
        {
            if(res !== undefined){
                if(res.userName){
                    dispatch(setUser(res))
                    AsyncStorage.setItem('session', JSON.stringify(res)) 
                }else{
                    dispatch(errorLoadingUser()); 
                }
            }
        });
    };
}

export const loadUser = () => {
    return function (dispatch) {
        AsyncStorage.getItem('session').then((session) => {
            if (session) {
                dispatch(setUser(JSON.parse(session)))
            }
        })
        dispatch(initCompleted())
    };
}

export const logOut = () => {
    return function (dispatch,getState) {
        FabricTwitterKit.logOut() //Doesnt work. Issue: https://github.com/tkporter/react-native-fabric-twitterkit/issues/33
        AsyncStorage.removeItem('session')
        dispatch(logOutAction())
    };
}

export const fetchProfile= (api = FabricTwitterKit) => {
    return function (dispatch) {
        api.fetchProfile((error, profile) => 
            {
                if(profile !== undefined){
                    dispatch(setProfile(profile))
                }
            })
    };
}

export const setProfile = (profile) => ({
    type: actionTypes.SET_PROFILE,
    name: profile.name,
    imageURL: profile.profile_image_url_https
})

export const logOutAction = () => ({
    type: actionTypes.LOG_OUT,
})

export const errorLoadingUser = () => ({
  type: actionTypes.ERROR_USER_LOGIN,
  error:true
})

export const initCompleted = () => ({
  type: actionTypes.INIT_COMPLETED,
  init:true
})

