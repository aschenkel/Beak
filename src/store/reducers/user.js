import * as actionTypes from '../actions/actionTypes'

var initialState = {
    init:false,
    authorized: false,
    userName: '',
    userID: '',
    authToken: '',
    authTokenSecret: '',
    error: false
}

const user = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.INIT_COMPLETED:
      return {
        ...state,
        init:true
      }
    case actionTypes.ERROR_USER_LOGIN:
      return {
        ...state,
        error:true
      }
    case actionTypes.LOAD_USER:
      return {
        ...state,
        userName: action.userName,
        userID: action.userID,
        authorized:true,
        authToken: action.authToken,
        authTokenSecret: action.authTokenSecret,
        error:false
      }
    default:
      return state
  }
}

export default user