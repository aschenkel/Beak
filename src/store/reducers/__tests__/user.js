import { initialState,user } from '../user';
import { setUser,setProfile,logOutAction,initCompleted } from '../../actions/user';

it('SetUser: should update state setting new user', () => {
  var userData= {
        userName: "test",
        userID: 0,
        authToken: 0,
        authTokenSecret: 0
    }
  const newState = user(initialState, setUser(userData));
  expect(newState).toMatchSnapshot();
});

it('SetProfile: should update state setting user profile', () => {
  var profileData= {
        userName: "test",
        imageURL: "https:/test"
    }
  const newState = user(initialState, setProfile(profileData));
  expect(newState).toMatchSnapshot();
});

it('LogOut: should update state to initialState with init to true', () => {
    var previousState= {
        authToken: "",
        authTokenSecret: "",
        authorized: true,
        error: false,
        imageURL: "https:/test",
        init: true,
        name: "test",
        tweets: [],
        userID: 0,
        userName: "@tester",
    }
    const newState = user(previousState, logOutAction());
    expect(newState).toEqual({
        ...initialState,
        init:true
     });
});

it('InitCompleted: should update state setting init to true', () => {
  const newState = user(initialState, initCompleted());
  expect(newState).toEqual({
        ...initialState,
        init:true
  });
});
