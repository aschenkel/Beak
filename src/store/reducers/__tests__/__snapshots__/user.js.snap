// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InitCompleted: should update state setting init to true 1`] = `
Object {
  "authToken": "",
  "authTokenSecret": "",
  "authorized": false,
  "error": false,
  "imageURL": null,
  "init": true,
  "name": "",
  "tweets": Array [],
  "userID": "",
  "userName": "",
}
`;

exports[`LogOut: should update state to initialState with init to true 1`] = `
Object {
  "authToken": "",
  "authTokenSecret": "",
  "authorized": false,
  "error": false,
  "imageURL": null,
  "init": true,
  "name": "",
  "tweets": Array [],
  "userID": "",
  "userName": "",
}
`;

exports[`SetProfile: should update state setting user profile 1`] = `
Object {
  "authToken": "",
  "authTokenSecret": "",
  "authorized": false,
  "error": false,
  "imageURL": undefined,
  "init": false,
  "name": undefined,
  "tweets": Array [],
  "userID": "",
  "userName": "",
}
`;

exports[`SetUser: should update state setting new user 1`] = `
Object {
  "authToken": 0,
  "authTokenSecret": 0,
  "authorized": true,
  "error": false,
  "imageURL": "",
  "init": false,
  "name": "",
  "tweets": Array [],
  "userID": 0,
  "userName": "test",
}
`;
